imbi/endpoints/project_ssm_configuration.py

```diff
@@ -0,0 +1,211 @@
+
+        output = [{
+            'name': name,
+            'type': data['type'],
+            'values': [{
+                'environment': environment,
+                'value': value
+            } for environment, value in data['values'].items()]
+        } for name, data in params_by_path.items()]
+        self.send_response(output)
+
+    async def get_aws_credentials(
+            self, role_arn, imbi_user: user.User, id_token,
+            refresh_token) -> tuple[bool, typing.Union[dict, str]]:
+        """Return whether the fetch succeeded, and the data or fail message"""
+        key = f'{imbi_user.username}:{role_arn}'
+        cached_creds = await self._redis.hgetall(key)
+        if cached_creds:
+            return True, {
+                key.decode('utf-8'): value.decode('utf-8')
+                for key, value in cached_creds.items()
+            }
+        try:
+            response = aws.get_credentials(role_arn, imbi_user.username,
+                                           id_token)
+        except botocore.exceptions.ClientError as error:
+            if error.response['Error']['Code'] == 'ExpiredTokenException':
+                id_token = await self.refresh_oauth2_tokens(
+                    imbi_user.username, imbi_user.external_id, refresh_token)
+                return await self.get_aws_credentials(role_arn, imbi_user,
+                                                      id_token, refresh_token)
+            else:
+                return False, error.response['Error']['Code']
+
+        creds = {
+            'access_key_id': response['Credentials']['AccessKeyId'],
+            'secret_access_key': response['Credentials']['SecretAccessKey'],
+            'session_token': response['Credentials']['SessionToken']
+        }
+        await self._redis.hmset_dict(key, creds)
+        await self._redis.expireat(
+            key, response['Credentials']['Expiration'].timestamp())
+        return True, creds
+
+    async def refresh_oauth2_tokens(self, username, external_id,
+                                    refresh_token) -> str:
+        env = os.environ.get('ENVIRONMENT', 'production').lower()
+        google_integration = await oauth2.OAuth2Integration.by_name(
+            self.application, f'google-{env}')
+        body = urllib.parse.urlencode({
+            'client_id': google_integration.client_id,
+            'client_secret': google_integration.client_secret,
+            'grant_type': 'refresh_token',
+            'redirect_uri': google_integration.callback_url,
+            'refresh_token': refresh_token,
+        })
+        response = await self.http_fetch(
+            str(google_integration.token_endpoint),
+            method='POST',
+            body=body,
+            content_type='application/x-www-form-urlencoded')
+        if response.code in (401, 403):
+            await self.session.clear()
+            await self.finish()
+            raise web.Finish()
```

Comment:
The `self.finish()` call is unnecessary unless we override [`_handle_request_exception()` in a weird and broken way](https://github.com/tornadoweb/tornado/blob/464f6084a5e18a01b46f77c011b1bbffe1f4d96c/tornado/web.py#L1833-L1837).